### DISCUSSION TITLE
Detect the mod

### DISCUSSION THREAD

**Discussion by TobiasHempel** [MinecraftFreecam/Freecam#261] [General]
Good evening,

I have a question about the mod from a server owner's point of view: Is there a way to check if a player is using this mod? We've been told by players that a few other players are using this mod, and it's actually not allowed on our server because it gives them an advantage in playing. Unfortunately, the mod is not recognized by the plugins we have tested. Are there any signs of how we can detect it? Basically there is nothing wrong with your mod, but it is very unfair on our survival servers.

I would be very happy about an answer and wish you a nice Christmas time.

**Comment by MattSturgeon** [comment]
> Is there a way to check if a player is using this mod?

Because the mod only affects the client and does not interact with the server in any way, it is very difficult (if not impossible) to detect usage server-side.

Theoretically, we could send some notification to the server that a user has activated Freecam and/or accept some configuration _from_ a server to enable disable particular settings and/or the mod entirely.

I don't think the first option would really gain much, and IMO isn't worth the effort for us to implement it. It wouldn't give server owners much reassurance either, given users could just use a different freecam mod.

I think there's some merit to allowing servers to configure config restrictions, although it could also be bypassed by users using a different mod it would at least allow _some_ servers to say "freecam mods that support server-set config restrictions are allowed". _Enforcing_ that would of course still be very difficult.

The server-set restrictions idea is tracked at #105

**Reply by MattSturgeon** [reply]
As an aside, it may be possible to use an exploit to detect if freecam is installed, e.g. the one mentioned in our README.


> While Freecam should only affect the client, it **may be possible** for servers to detect the mod being installed through various exploits in Minecraft, modding platforms, or this mod.  
> Here's a [real example](https://github.com/MinecraftFreecam/Freecam/issues/196) that exploited translation keys.

Knowing a mod is installed _does not_ tell you it was actually used on your server though. Therefore I don't recommend using it to ban players.

**Reply by TobiasHempel** [reply]
Good evening. Thank you very much for your reply!
It's exactly my intention that we first find out whether what the other players have said is true. We don't want to simply ban anyone, but it would still be important for us to know for sure whether they are using the plugin.
I understand your worries that players could simply use other Freecam mods if this plugin sends secret messages to the server. That's why I wanted to try it now with this exploid.
Unfortunately, I haven't programmed Minecraft in a few years. Do you know if there is already a plugin that tests the exploit? I just tried it, but so far I can only get it to display different messages to players on signs. However, when I print out the sign lines, it only says (for example) “msg.freecam.enable”. The client translates it to freecam has... (sign ends)
Do you have any idea how to catch this and see from the server whether the client replaces the message or not? or do you have any other ideas?
Maybe my request doesn't fit in here. But I think it could also be interesting for others. Thanks again for your efforts!

**Reply by UrTheProblem** [reply]
I would suggest to maybe add a registration of a plugin channel, which other mods like tweakeroo or xaeors minimap also do. This makes it easier to detect with a plugin and also makes it so the plugin doesn't need to rely on exploits to detect the mod

**Reply by MattSturgeon** [reply]
Are you willing to put together a proof or concept PR? Or point to a working example? I don't currently have time to dig through tweakeroo's codebase to see if/how they do this, but I'm open to the idea.

We also have related issues, such as https://github.com/MinecraftFreecam/Freecam/issues/105 and https://github.com/MinecraftFreecam/Freecam/issues/329, that discuss creating an API for servers to have a say over Freecam's config. That's a much more involved idea, but maybe registering a "server plugin" could form the basis of that?

**Reply by UrTheProblem** [reply]
Sorry for the late reply,

I haven’t worked with pluginchannels on the client side yet sadly. They are registered on join and can be used by both mod and plugin to communicate. For example mods like simple voice chat use pluginchannels and xaeros maps use them for server handeled settings aswell.

I‘ll suggest this approach in the related issues, but i sadly am not able to make a PR, due to a lack of time and knowledge as i have only discovered them recently
